### ISSUE TITLE
Resolving conflicts between node and local header/src includes.

### ISSUE THREAD

**Issue by royalpinto** [nodejs/node-gyp#988]
Hi,

**I want to ensure that node-gyp always considers headers files from local project than from `node` if there is conflict.**

This is in continuation with the issue [https://github.com/nodejs/node-gyp/issues/986](https://github.com/nodejs/node-gyp/issues/986).

**The project:**
[node-cares](https://github.com/royalpinto/node-cares) makes use of `c-ares` lib and is similar to default `node` `dns` module but with additional features like creating multiple instances of `Resolver`s with various configs.

**The problem**:
I have `deps/` dir within the project in which `c-ares` source code is copied. Now I would like ensure that `node-gyp` always considers src and headers files from within the project directory and not from `node`.

The binding.gyp:
```
"include_dirs": [
    "<!(node -e \"require('nan')\")",
    "deps/cares/include",
    "deps/cares/src"
],
"sources": [
    # I can ensure that this source is always considering local `cares` copy
    # by setting relative path in the #includes
    # Ex: #include "../deps/cares/include/ares.h"
    "src/cares_wrap.cc"
],
# The problem is here when this dependency is getting build.
# It always refers to the `node` source files than local copy.
# I would like to ensure that this will refer to the local copy.
"dependencies": [ "deps/cares/cares.gyp:cares" ],
"conditions": [
    ["OS!='win'", {
          "libraries": [ "-Wl,-rpath,<!(pwd)/build/Release/" ]
        }
    ]
]
```

Now that node has it's own version of `cares` and differ from version to version and also base on the discussion [here](https://github.com/nodejs/node-gyp/issues/986), I can't really refer to the `node` code and local copy of `cares` source is must. When I try to build the project, it always refers to the `node`s `cares` than referring to the local copy for `dependencies:`.

Here is the .target.mk file which gets created by node-gyp (Note: local copies are at the end):
```
INCS_Release := \
	-I/Users/lpinto/.node-gyp/4.3.0/include/node \
	-I/Users/lpinto/.node-gyp/4.3.0/src \ ###I want to remove this or keep it at the end.
	-I/Users/lpinto/.node-gyp/4.3.0/deps/uv/include \
	-I/Users/lpinto/.node-gyp/4.3.0/deps/v8/include \
	-I$(srcdir)/node_modules/nan \
	-I$(srcdir)/deps/cares/include \
	-I$(srcdir)/deps/cares/src
```
Is there any way, I can enforce `node-gyp` to consider my local copy on priority before considering `cares` source from `node` ? With earlier versions of `node-gyp` and `node`, this was not the problem because .mk used to get create without the root `node` dir `-I/Users/lpinto/.node-gyp/4.3.0/src \` and used to always refer to local copy.

If its not possible and every source file within `cares` I have to modify to refer to the local copy by setting the relative path :(

Thanks

**Comment by rvagg** [comment]
OK, this is super hacky but it should work. We might have to get @bnoordhuis to give a 👍 on whether this is actually safe though.

Replace your `"include_dirs"` block with these two:

```
"include_dirs!": [
    "<(node_root_dir)/include/node",
    "<(node_root_dir)/src",
    "<(node_root_dir)/deps/uv/include",
    "<(node_root_dir)/deps/v8/include"
],
"include_dirs": [
    "<!(node -e \"require('nan')\")",
    "deps/cares/include",
    "deps/cares/src",
    "<(node_root_dir)/./include/node",
    "<(node_root_dir)/./src",
    "<(node_root_dir)/./deps/uv/include",
    "<(node_root_dir)/./deps/v8/include"
],
```

The first block removes the defaults from [addon.gypi](https://github.com/nodejs/node-gyp/blob/master/addon.gypi) and then in the second block we put them back again as they should be but the `'./` addition in the middle stops them from being matched from the first so they won't be removed, they end up being resolved properly by GYP in the end, however.

The main problem with this approach is that if we ever make changes in addon.gypi then you won't inherit them in your project, but that shouldn't happen very often, if at all.

**Comment by royalpinto** [comment]
Hi @rvagg,

Wow that was awesome! 👍, thanks a lot.  Let me give this a try.
If `addon.gypi` doesn't get modified often, I can manage to update my `binding.gyp` whenever there is a change in `addon.gypi`.

Hi @bnoordhuis, do you see any risk involved in this solution ?

Thank you  :)

**Comment by bnoordhuis** [comment]
What @rvagg suggests should work (but warranty void if used, obviously.)

You can perhaps make it a little more robust by moving cares_wrap.cc to a separate static_library target, then you don't have to re-add the default includes.

**Comment by royalpinto** [comment]
Thanks @bnoordhuis 👍 

Will try this `static_library` target thing as well.

Thanks again!

**Comment by royalpinto** [comment]
Hi @rvagg, @bnoordhuis,

`include_dirs!` worked for me. I had to exclude default includes, only in `cares.gyp` which was anyway not using default includes, re-add also wasn't required. In the main source file `cares_wrap.cc` I had to modify just two lines to refer to the local `cares` than to the `node`.

Here are the changes made to cares_wrap.cc:
```
-#include "ares.h"
-#include "ares_dns.h"
+#include "../deps/cares/include/ares.h"
+#include "../deps/cares/src/ares_dns.h"
```

And cares.gyp:
```
+      'include_dirs!': [
+         '<(node_root_dir)/include/node',
+         '<(node_root_dir)/src',
+         '<(node_root_dir)/deps/uv/include',
+         '<(node_root_dir)/deps/v8/include'
+      ],
```

Only one thing I would have to take care is, if `addon.gypi` includes new dir (which is not very often) and if its conflicting local headers, exclude them in `cares.gyp` which is easily manageable.

Thanks a lot 👍
